### PR TITLE
Fix/tf idf fit transform

### LINKED ISSUE #345: Tf-idf vectoriser is lemmatising twice
The transformations should only be in "transform" not in "fit".

Atm we have:

```bash
2021-07-29 13:02:16 wellcomeml.ml.frequency_vectorizer INFO: Using spacy pre-trained lemmatiser.
2021-07-29 13:03:15 wellcomeml.ml.frequency_vectorizer INFO: Fitting vectorizer.
2021-07-29 13:03:24 wellcomeml.ml.frequency_vectorizer INFO: Using spacy pre-trained lemmatiser.
```

When we run fit-transform this is causing problems because it's double lemmatising.

### PR DESCRIPTION
**Pull request by aCampello**
Description
---

Fixes #345 

Also improves performance a bit by initialising spacy open initialisation of class.

Checklist
---

- [x] Added link to Github issue or Trello card
- [ ] Added tests

### CODE REVIEW

**Comment by nsorros** [comment]
It would be great to add a test to demonstrate the fix i.e. a fix that would fail prior to fixing this

File: wellcomeml/ml/frequency_vectorizer.py [hunk 1]
```diff
@@ -115,16 +115,9 @@ def transform(self, X, regex=True, spacy_lemmatizer=True, *_):
...
-
-        X = [" ".join(text) for text in X]
-
-        super().fit(X)
-
+    def fit(self, X, *_):
+        super().fit()
         return self
+
+    def fit_transform(self, X, y=None):
```
Comment (nsorros) [inline wellcomeml/ml/frequency_vectorizer.py:122]: isn't this provided for free from sklearn?

**Comment by nsorros** [comment]
How does this prevent the `lemmatiser` running twice?

**Comment by aCampello** [comment]
> How does this prevent the `lemmatiser` running twice? Is it because it does not run during `fit`? If so then that sounds wrong

The fit function of TfIdf is stateless, it does nothing. All the heavy-lifting is done in `.transform`. So any transformations should only run in `.transform`, but we were wrongly putting it into `.fix`. This doesn't interfere that much, but it was causing weird errors like, if you run `X_transform = vectorizer.fit_transform(X)` and then `X_transform_2 = vectoriser.fit(X)` you wouldn't end up with the same results. 

I will add a test to demonstrate that and also fix the test.

**Comment by aCampello** [comment]
Check: 

https://github.com/scikit-learn/scikit-learn/blob/2beed55847ee70d363bdbfe14ee4401438fba057/sklearn/feature_extraction/text.py#L720-L768

**Comment by aCampello** [comment]
Thanks for the test suggestion @nsorros . 

So, I added a test to check that `.fit(X).transform(X)` was the same as `fit_transform(X)`. Without the new code it fails:

```bash
        # Asserts that the result of transform is almost the same as fit transform
>       assert (X_embed-X_embed_2).sum() == pytest.approx(0, abs=1e-6)
E       assert 0.18267581368159946 == 0 ± 1.0e-06
E         +0.18267581368159946
E         -0 ± 1.0e-06

tests/test_frequency_vectorizer.py:47: AssertionError
```

With the new code it passes.

Check https://github.com/wellcometrust/WellcomeML/pull/346/commits/d272de193a79aec71525e5be312f8315d55fa3bb

File: wellcomeml/ml/frequency_vectorizer.py (hunk 1 above)
Comment (aCampello) [reply wellcomeml/ml/frequency_vectorizer.py:122]: That's why the `TfId` class in sklearn has to implement their own `.fit.` even though it's a transform mixin.

Comment (nsorros) [reply wellcomeml/ml/frequency_vectorizer.py:122]: 👍

**Comment by nsorros** [comment]
> Check:
> 
> https://github.com/scikit-learn/scikit-learn/blob/2beed55847ee70d363bdbfe14ee4401438fba057/sklearn/feature_extraction/text.py#L720-L768

This is the `fit` of `HashingVectorizer` which is indeed stateless. TfidfVectorizer is not `stateless`, it updates the vocabulary see https://github.com/scikit-learn/scikit-learn/blob/2beed55847ee70d363bdbfe14ee4401438fba057/sklearn/feature_extraction/text.py#L1806

**Comment by nsorros** [comment]
> > How does this prevent the `lemmatiser` running twice? Is it because it does not run during `fit`? If so then that sounds wrong
> 
> The fit function of TfIdf is stateless, it does nothing. All the heavy-lifting is done in `.transform`. So any transformations should only run in `.transform`, but we were wrongly putting it into `.fix` too. This doesn't interfere that much because "doubly" lemmatising something should be almost idempotent, but it was causing weird errors like, if you run `X_transform = vectorizer.fit_transform(X)` and then `X_transform_2 = vectorizer.fit(X)` you wouldn't end up with the same results.
> 
> I will add a test to demonstrate that and also fix the test.

I think this is wrong. `fit` calculates the vocabulary and you were correctly lemmatising before doing so, you should lemmatise both for fit and transform as they do different things, kind of like preprocessing, you need to preprocess with the same function.

I think what was potentially going wrong is that `fit_transform` passes the X to both fit and transform but you were modifying X in `fit`, you need to save in a different variable before calling the `super().fit`

**Review by nsorros — CHANGES_REQUESTED** [review]
I think introduces a problem as the vocabulary is calculated on the non lemmatised text whereas transform lemmatises. This can lead to unexpected results.

**Comment by aCampello** [comment]
> > Check:
> > https://github.com/scikit-learn/scikit-learn/blob/2beed55847ee70d363bdbfe14ee4401438fba057/sklearn/feature_extraction/text.py#L720-L768
> 
> This is the `fit` of `HashingVectorizer` which is indeed stateless. TfidfVectorizer is not `stateless`, it updates the vocabulary see https://github.com/scikit-learn/scikit-learn/blob/2beed55847ee70d363bdbfe14ee4401438fba057/sklearn/feature_extraction/text.py#L1806

You're right, the link was wrong. The TfIdf class is actually very clunky, all the heavy-lifting is done in `fit_transform` and then the `fit` just calls `fit_transform`. 

I'll modify it to look more like TfIdf, thanks for the review.

**Review by nsorros — APPROVED** [review]
LGTM 👍

**Comment by aCampello** [comment]
Ok, merging this guy, thanks for spotting the error @nsorros !

**Pull request merged**